In [4]:
import pandas as pd

# Dataset Information
### This dataset contains the data of 10 cars in the CARLA simulation environment, with the features as shown below.

In [5]:
df = pd.read_csv('vehicle_data.csv')

In [6]:
df.head()

timestamp  vehicle_id  gear  throttle     speed  acceleration  max_rpm  \
0  1.708225e+09           0     0      0.85  1.061667     25.723619   6300.0   
1  1.708225e+09           1     0      0.85  0.816886     28.389420   4700.0   
2  1.708225e+09           0     0      0.85  0.274378      0.511369   6300.0   
3  1.708225e+09           1     0      0.85  0.239636      0.081476   4700.0   
4  1.708225e+09           0     0      0.85  0.076758      0.820615   6300.0   

   gear_ratio  wheel_radius  engine_rpm  inclination tyre_pressure  odometer  \
0       3.450          0.33   30.737318     0.068042  (0, 0, 0, 0)  0.000000   
1       3.769          0.34   22.954825     0.000000  (0, 0, 0, 0)  0.000000   
2       3.450          0.33    7.943773     0.136481  (0, 0, 0, 0)  0.029130   
3       3.769          0.34    6.733879     0.000000  (0, 0, 0, 0)  0.006514   
4       3.450          0.33    2.222298     0.082611  (0, 0, 0, 0)  0.062866   

   fuel_consumed  ignition_status  engine_oil_life  new_battery_percent  
0   0.000000e+00             True       100.000000            99.998069  
1   0.000000e+00             True       100.000000            99.997473  
2   5.490478e-06             True        99.999999            99.996138  
3   4.736946e-07             True       100.000000            99.994946  
4   1.507606e-05             True        99.999999            99.994207

In [7]:
df = df[df["vehicle_id"] == 0]

In [8]:
df.head()

timestamp  vehicle_id  gear  throttle     speed  acceleration  max_rpm  \
0  1.708225e+09           0     0      0.85  1.061667     25.723619   6300.0   
2  1.708225e+09           0     0      0.85  0.274378      0.511369   6300.0   
4  1.708225e+09           0     0      0.85  0.076758      0.820615   6300.0   
6  1.708225e+09           0     1      0.85  1.681097      5.602935   6300.0   
8  1.708225e+09           0     1      0.85  2.274781      2.170914   6300.0   

   gear_ratio  wheel_radius  engine_rpm  inclination tyre_pressure  odometer  \
0        3.45          0.33   30.737318     0.068042  (0, 0, 0, 0)  0.000000   
2        3.45          0.33    7.943773     0.136481  (0, 0, 0, 0)  0.029130   
4        3.45          0.33    2.222298     0.082611  (0, 0, 0, 0)  0.062866   
6        2.10          0.33  401.681974     0.610148  (0, 0, 0, 0)  0.251549   
8        2.10          0.33  543.537085     0.467403  (0, 0, 0, 0)  0.658497   

   fuel_consumed  ignition_status  engine_oil_life  new_battery_percent  
0       0.000000             True       100.000000            99.998069  
2       0.000005             True        99.999999            99.996138  
4       0.000015             True        99.999999            99.994207  
6       0.000055             True        99.999994            99.992276  
8       0.000090             True        99.999985            99.990345

In [10]:
X = df[["speed", "throttle", "acceleration", "max_rpm"]]

In [11]:
y = df["engine_rpm"]

In [12]:
from sklearn.model_selection import train_test_split

In [13]:
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2)

In [14]:
from sklearn.ensemble import RandomForestRegressor

In [15]:
rfc = RandomForestRegressor(n_estimators=100)

In [16]:
rfc.fit(X_train, y_train)

RandomForestRegressor()

In [17]:
predictions = rfc.predict(X_test)

In [18]:
from sklearn.metrics import mean_absolute_error, mean_squared_error

In [19]:
mean_absolute_error(y_test, predictions)    

15.80252421149909

In [21]:
import numpy as np

In [24]:
rmse = np.sqrt(mean_squared_error(y_test, predictions))

In [27]:
print("RMSE:", rmse)

RMSE: 68.57361840631863


In [29]:
print("Accuracy:",rfc.score(X_test, y_test))

Accuracy: 0.9835715187545178


# Conclusion
### We have trained a RandomForestRegressor to predict the engine RPM with 98% accuracy.